### PR TITLE
Feature: Add support for JUnit 5 tests via HdfsExtension.

### PR DESCRIPTION
**Pull request by mojtab23**

### CODE REVIEW

File: src/main/java/ir/sahab/hdfsrule/HdfsBase.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+class HdfsBase {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hdfsrule/HdfsBase.java]: Add Java Doc

File: src/main/java/ir/sahab/hdfsrule/HdfsBase.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            conf.set(MiniDFSCluster.HDFS_MINIDFS_BASEDIR, miniClusterDataDir.getAbsolutePath());
+            hdfsCluster = new MiniDFSCluster.Builder(conf).build();
+            hdfsCluster.waitActive();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    void teardown() {
+        lock.lock();
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hdfsrule/HdfsBase.java:47]: Explain why we need this locking mechanism

File: src/main/java/ir/sahab/hdfsrule/HdfsExtension.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package ir.sahab.hdfsrule;
+
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.extension.*;
+
+/**
+ * A JUnit 5 extension for starting an HDFS server on the local machine.
+ */
+public class HdfsExtension extends HdfsBase implements BeforeAllCallback, AfterAllCallback, BeforeEachCallback, AfterEachCallback {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hdfsrule/HdfsExtension.java:12]: Also explain the PER_CLASS and PER_METHOD use cases and default behavior in java doc.

File: src/main/java/ir/sahab/hdfsrule/HdfsRule.java [hunk 4]
```diff
@@ -1,86 +1,54 @@
...
-import org.apache.hadoop.hdfs.MiniDFSCluster;
+
+import java.io.IOException;
+import java.net.URI;
 
 /**
- * A JUnit rule for starting an HDFS server on the local machine.
+ * A JUnit 4 rule for starting an HDFS server on the local machine.
  */
 public class HdfsRule extends ExternalResource {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hdfsrule/HdfsRule.java]: public class HdfsRule extends HdfsBase implements TestRule {

File: src/main/java/ir/sahab/hdfsrule/HdfsBase.java (hunk 2 above)
Comment (mojtab23) [reply src/main/java/ir/sahab/hdfsrule/HdfsBase.java:47]: You said to use the lock, maybe for concurrent tests.

File: src/main/java/ir/sahab/hdfsrule/HdfsRule.java (hunk 4 above)
Comment (mojtab23) [reply src/main/java/ir/sahab/hdfsrule/HdfsRule.java]: I can't the before and after methods that we override are coming from ExternalResource.

**Pull request merged**